packages/flutter_test/lib/src/controller.dart

```diff
@@ -0,0 +1,277 @@
+  Iterable<Widget> get allWidgets {
+    return allElements
+           .map((Element element) => element.widget);
+  }
+
+  /// The matching widget in the widget tree.
+  ///
+  /// Throws a [StateError] if `finder` is empty or matches more than
+  /// one widget.
+  Widget/*=T*/ widget/*<T extends Widget>*/(Finder finder) {
+    return finder.evaluate().single.widget;
+  }
+
+  /// The first matching widget according to a depth-first pre-order
+  /// traversal of the widget tree.
+  ///
+  /// Throws a [StateError] if `finder` is empty.
+  Widget/*=T*/ firstWidget/*<T extends Widget>*/(Finder finder) {
+    return finder.evaluate().first.widget;
+  }
+
+  /// All elements currently in the widget tree (lazy pre-order traversal).
+  ///
+  /// The returned iterable is lazy. It does not walk the entire widget tree
+  /// immediately, but rather a chunk at a time as the iteration progresses
+  /// using [Iterator.moveNext].
+  Iterable<Element> get allElements {
+    return collectAllElementsFrom(binding.renderViewElement);
+  }
+
+  /// The matching element in the widget tree.
+  ///
+  /// Throws a [StateError] if `finder` is empty or matches more than
+  /// one element.
+  Element/*=T*/ element/*<T extends Element>*/(Finder finder) {
+    return finder.evaluate().single;
+  }
+
+  /// The first matching element according to a depth-first pre-order
+  /// traversal of the widget tree.
+  ///
+  /// Throws a [StateError] if `finder` is empty.
+  Element/*=T*/ firstElement/*<T extends Element>*/(Finder finder) {
+    return finder.evaluate().first;
+  }
+
+  /// All states currently in the widget tree (lazy pre-order traversal).
+  ///
+  /// The returned iterable is lazy. It does not walk the entire widget tree
+  /// immediately, but rather a chunk at a time as the iteration progresses
+  /// using [Iterator.moveNext].
+  Iterable<State> get allStates {
+    return allElements
+           .where((Element element) => element is StatefulElement)
+           .map((StatefulElement element) => element.state);
+  }
+
+  /// The matching state in the widget tree.
+  ///
+  /// Throws a [StateError] if `finder` is empty, matches more than
+  /// one state, or matches a widget that has no state.
+  State/*=T*/ state/*<T extends State>*/(Finder finder) {
+    return _stateOf/*<T>*/(finder.evaluate().single, finder);
+  }
+
+  /// The first matching state according to a depth-first pre-order
+  /// traversal of the widget tree.
+  ///
+  /// Throws a [StateError] if `finder` is empty or if the first
+  /// matching widget has no state.
+  State/*=T*/ firstState/*<T extends State>*/(Finder finder) {
+    return _stateOf/*<T>*/(finder.evaluate().first, finder);
+  }
+
+  State/*=T*/ _stateOf/*<T extends State>*/(Element element, Finder finder) {
+    if (element is StatefulElement)
+      return element.state;
+    throw new StateError('Widget of type ${element.widget.runtimeType} found by ${finder.description} does not correspond to a StatefulWidget.');
```

Comment:
Changed to `'Widget of type ${element.widget.runtimeType}, with ${finder.description}, is not a StatefulWidget.'`